### ISSUE TITLE
[Due for payment 2025-03-10] [$250] Expense status doesn't update after completing the action

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#46914]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.17-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @saracouto 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1722617920576379

## Action Performed:
1. Submit an expense and get a violation
2. Fix the violation
3. Violation message doesn't update in the upper side of the screen

## Expected Result:
Once the user clears the violation, the message should update from "Waiting for Sara Couto to fix the issue(s)" to the corresponding next step
## Actual Result:
The message "Waiting for Sara Couto to fix the issue(s)" was still there even after fixing the violation
## Workaround:
unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos
![image (106)](https://github.com/user-attachments/assets/ea901700-63b5-4f24-8c1f-043a3358a956)

https://github.com/user-attachments/assets/013b1346-7811-4300-a4f7-2d22f153a80d

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01bfe746236df54daa</li>
        <li>Upwork Job ID: 1821182632521988875</li>
        <li>Last Price Increase: 2024-08-14</li>
<li>Automatic offers: </li>
<ul>
<li>rayane-djouah | Reviewer | 103528411</li>
<li>dominictb | Contributor | 103528413</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @slafortune</details>

**Comment by trjExpensify** [comment]
Moving to #wave-control as it's optimistic next steps/violations related. CC: @dangrous @dylanexpensify

**Comment by dangrous** [comment]
assigning myself so I can take a look when proposals are approved!

**Comment by slafortune** [comment]
Adding another BZ - I'll be out until 8/21

**Comment by dangrous** [comment]
Yep, looks good to me! I'm wondering - to solve the "There's no violation before, but after edit there's violation" case, could we just do: `currentTransactionViolations.length != nextViolationOnyxUpdate.value.length`?

**Comment by dylanexpensify** [comment]
Nice, @dominictb when can we get the PR raised for review?

**Comment by dangrous** [comment]
Okay so in talking with @mountiny and trying to make this work on the backend, it's going to bog down our code speed and also just be very tricky to do it currently. However! We're in the process of migrating the next steps code, and once that's done we should be able to fix this much more easily.

It does mean in the meantime, though, that we won't want to merge this front-end fix if it will just be overwritten by the backend. So I think this (and the PR, which I'll post on too) will be on hold for the next steps migration to auth. Sorry for the trouble!

**Comment by dangrous** [comment]
Still on hold Melvin, don't close this.

**Comment by dangrous** [comment]
We've actually got a temporary fix that should get this moving again, should be merged soon. We're still doing the migration, but I'm pretty sure this will work in the meantime.

**Comment by dangrous** [comment]
Okay I thinkkkkk this is deployed. Can we retest and see if it's fixed?

**Comment by dangrous** [comment]
hrm can you let me know how you set up the workspace / rules / etc. so I can try to replicate? I've found one place in the code that I can fix, but I think there's another one. 

Also @mountiny did we figure out how to fix that issue with the cached violations? That's not this bug, but it's related.

**Comment by mountiny** [comment]
> Also @mountiny did we figure out how to fix that issue with the cached violations? That's not this bug, but it's related.

I think yes, will send you a PR

**Comment by mountiny** [comment]
This is the PR btw https://github.com/Expensify/Web-Expensify/pull/44311 however, after testing it locally the cache seems to not get updated  before the report next steps are computed. It will require more digging 🕳️

**Comment by dangrous** [comment]
Thanks @Dominic! So we're looking at two next step issues (cc @mountiny).

For a policy on Instant Submit, we're hitting [this block](https://github.com/Expensify/Web-Expensify/blob/6564b3b024def86cb423e5c0ffb556671b93cb61/lib/Report/NextStep.php#L328-L337) even if a report has violations. We should probably add a condition there, that's relatively easy, and doesn't change any existing tests.

For a policy on manual submit, we SHOULD be hitting [this block](https://github.com/Expensify/Web-Expensify/blob/6564b3b024def86cb423e5c0ffb556671b93cb61/lib/Report/NextStep.php#L415-L426), but we're not because the logic is checking a lot of stuff about that specific user, and it's conflicting for some reason. So instead we're going down [here](https://github.com/Expensify/Web-Expensify/blob/6564b3b024def86cb423e5c0ffb556671b93cb61/lib/Report/NextStep.php#L549-L556). I think we can probably just drop everything after `$hasViolations` since everything is more streamlined now? But I'm not 100% sure. The tests would need some adjustment but I'm hopelessly confused now as to what is "expected".

**Comment by mountiny** [comment]
@dangrous sounds right, but I agree this is really finicky and gotta be careful around the test

**Comment by dangrous** [comment]
So I think a lot of the tests are messed up, annoyingly. E.g they're creating reports with expenses that have violations, when that's not the behavior they're testing. See my latest comment on https://github.com/Expensify/Web-Expensify/pull/44527 . I'm trying to power through and make these changes correctly, but a little stuck

**Comment by dangrous** [comment]
okay so the caching is fixed, the instant submit case is fixed, I just need to implement the manual submit case. And then we should be good?

**Comment by dangrous** [comment]
okay I think I got all of it in that PR - will take it out from draft momentarily

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by dangrous** [comment]
Amazing! Where do we stand on this issue, all? Assuming we can't reproduce anymore, can we just close? Is payment needed? Further front end work? Thanks!

**Comment by dangrous** [comment]
this is on prod now, so bumping ^ - thanks!

**Comment by dangrous** [comment]
okay well we just reverted the PR haha so that's your issue here... I think we're back to the drawing board unfortunately. We'll be taking another look at the backend today. One day we'll get there.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by dangrous** [comment]
New version of the fix for this is now in backend review!

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Third week)

**Comment by dangrous** [comment]
Reopening since we're still working on it

**Comment by dangrous** [comment]
And just to have a more useful update, we're still discussing on the backend the best way to go about this with minimal performance hits. I (egotistically) think I have the best answer and am trying to convince others of it :). Hopefully we'll resolve on something soon!

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Fourth week)

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Fifth week)

**Comment by dangrous** [comment]
hey! still waiting on final review for one, but it should be merged today or tomorrow. Then we will hopefully be good to go!

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Sixth week)

**Comment by dangrous** [comment]
Okay I think this is finally done on the backend. Can we retest and then close out if we're good to go?

**Comment by slafortune** [comment]
Whomp!  Sure - adding the retest label again 👍

**Comment by dangrous** [comment]
ugggh i don't know why this is not working anymore, this is the exact test I was using before....

But you're right it's not working now. Trying to figure it out. Basically it's because the internal `$hasViolations` is still returning false...

**Comment by dangrous** [comment]
okay so weird, so the violation is indeed there but it's marked as `showInReview` as `false` which removes it from the list. Not sure why that's happening now and wasn't before?

**Comment by dangrous** [comment]
hey @dominictb - can you confirm if you had a `description` for your expense? I think having a `description` negates the error (since you're "explaining" it) so it won't show up as the next step. That, at least, was what was happening on my testing - it works for me on manual submit with an error - as long as I don't have a description - both locally and on staging.

**Comment by dangrous** [comment]
hrm yeah I'm still getting the expected result on staging:

<img width="859" alt="Image" src="https://github.com/user-attachments/assets/2bae8743-fc4a-45b8-9614-0cbf4cb47055" />

My workspace setup:

<img width="851" alt="Image" src="https://github.com/user-attachments/assets/b80e574f-4a05-4ff8-a8a9-5bb76962a455" />

<img width="858" alt="Image" src="https://github.com/user-attachments/assets/77e7d063-71bb-4470-8150-ee7d8dcd0507" />

Anything you see that's different?

**Comment by dangrous** [comment]
great, thank you!

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Seventh week)

**Comment by dangrous** [comment]
Updated! Can you share a link to the PR again when you have a moment? I can't find it in the GH since there are so many comments haha

**Comment by slafortune** [comment]
@rayane-d Role C+ requires payment through NewDot Manual Requests owed $250
@dominictb Role Contributor paid via Upworks - $250 ✅
